File: leetcode/third_max_number.cpp
class Solution {
public:
    int thirdMax(vector<int> nums) {
        int max1=INT_MIN, max2=INT_MIN, max3=INT_MIN;
        bool max3_seen=false;

        for(int i=0; i< nums.size(); i++) {
            if(nums[i] > max1) {
                max3=max2;
                max2=max1;
                max1=nums[i];
            } else if(nums[i] > max2) {
                if(nums[i] != max1) {
                    if(max2 != max3) max3_seen = true;
                    max3=max2;
                    max2=nums[i];
                }
            } else if(nums[i] > max3) {
                if(nums[i] != max2) {
                    max3=nums[i];
                    max3_seen = true;
                }
            }
        }

        if(max3_seen) return max3;
        else return max1;
    }
};

File: leetcode/third_max_number.java
class Solution {
    public int thirdMax(int[] nums) {
        Integer m1=null,m2=null,m3=null;
        for (Integer n:nums){
            if (n.equals(m1) || n.equals(m2) || n.equals(m3)) continue;
            if (m1==null || n>m1){
                m3=m2;
                m2=m1;
                m1=n;
            }else if (m2==null || n>m2){
                m3=m2;
                m2=n;
            }else if (m3==null || n>m3) m3=n; 
        }
        return m3==null? m1: m3;
    }
}
